### DISCUSSION TITLE
`security import` command suddenly causing error in macOS runner

### DISCUSSION THREAD

**Discussion by mywristbands** [community/community#63731] [Actions] [answered]
### Select Topic Area

Question

### Body

I have a GitHub Actions workflow that deploys an iOS app to App Store Connect. One of the steps involves importing a certificate to the Keychain. My code to perform this importing step has been the same since I originally set up the workflow almost a year ago. Here is the code:
```
security import <(echo $SIGNING_CERTIFICATE_P12_DATA | base64 --decode) \
                -f pkcs12 \
                -k build.keychain \
                -P $SIGNING_CERTIFICATE_PASSWORD \
                -T /usr/bin/codesign
```

However, as of this past Thursday, August 10, 2023, this code now leads to the following errors:

```
security: SecKeychainItemImport: One or more parameters passed to a function were not valid.
bad decrypt
140704704464640:error:02012020:system library:fflush:Broken pipe:crypto/bio/bss_file.c:316:fflush()
140704704464640:error:20074002:BIO routines:file_ctrl:system lib:crypto/bio/bss_file.c:318:
```

I have tested this code on my local computer running the same version of macOS (macOS 13), and it works just fine. Note that this code is part of a larger script that also creates and configures the `build.keychain` keychain, so I tried running this entire script locally as well (I even used a bash terminal just like is used in GitHub Actions), and this script works locally just fine. So I'm not sure why it's not working in GitHub Actions.

Has there been some change to any of the commands I use in my importing code above that might have caused this to not work anymore? Does anyone have suggestions to fix it? If it would be helpful, I can provide the whole script involved in setting up the keychain and importing the certificate.

**Comment by aljux** [comment]
I ran into this issue as well. If you're using cocoapods in the build pipeline, try to update cocoapods to 1.12.1. Solved the problem for me. Not sure why though.

**Comment by oftharsh** [comment]
The error you're encountering, "One or more parameters passed to a function were not valid," along with the subsequent "bad decrypt" and "Broken pipe" errors, suggests that there might be an issue with the input data or the execution environment in GitHub Actions. Here are some steps you can take to troubleshoot and potentially resolve the problem:

1. **Verify Environment Variables**: Double-check that the environment variables like `SIGNING_CERTIFICATE_P12_DATA` and `SIGNING_CERTIFICATE_PASSWORD` are correctly set in your GitHub Actions workflow. Ensure there are no extra spaces, newline characters, or encoding issues.

2. **Base64 Encoding**: Confirm that the `SIGNING_CERTIFICATE_P12_DATA` variable is being properly base64 encoded when it's generated or stored. Make sure that the data is being decoded correctly in your workflow.

3. **Version Compatibility**: Verify that the version of `security` utility on GitHub Actions matches the one on your local machine. You can print the version using `security version`.

4. **Keychain**: Ensure that the `build.keychain` keychain is being created and configured correctly in your workflow before the `security import` step. Check for any potential issues with permissions or keychain settings.

5. **Pipe Handling**: There might be some subtle differences in how pipes and process substitution (`<()`) work in the GitHub Actions environment compared to your local shell. Consider using temporary files to pass data to `security import` instead of process substitution.

6. **Debugging**: Add debug output statements to your workflow to print out the values of relevant variables and command outputs at various steps. This can help you identify where the problem might be occurring.

7. **GitHub Actions Updates**: Check if there have been any updates or changes to the GitHub Actions environment, macOS version, or other relevant components that might be affecting the behavior of your script.

8. **GitHub Support**: If the issue persists, consider reaching out to GitHub Support for assistance. They might be able to provide more insights or help troubleshoot the issue further.

Remember to be cautious with handling sensitive information like certificates and passwords in your workflow, and consider using GitHub's Secrets feature to securely store and retrieve such data.

Please note that my knowledge is based on information available up to September 2021, and there might have been changes or updates since then.

**Comment by Jaber-Saed** [comment]
The error you're encountering during the certificate import step in your GitHub Actions workflow seems to be related to a problem with the security command when using a pipe to pass in data. The "One or more parameters passed to a function were not valid" message indicates that there might be an issue with the input data or the command execution.

Since this issue seems to have started recently, it's possible that there might have been changes or updates to the GitHub-hosted runner environments or macOS system libraries that are affecting the behavior of the security command. Here are a few steps you can try to troubleshoot and address the issue:

    Base64 Decoding Issue: Double-check the value of the SIGNING_CERTIFICATE_P12_DATA variable that you're passing into the security command. Make sure that the base64-encoded certificate data is correct and hasn't been corrupted or modified.

    Quoting and Escaping: When using shell scripts in GitHub Actions, sometimes special characters can cause issues. Make sure you're quoting and escaping variables correctly. For example, use double quotes around variables that may contain spaces or special characters.

    Debugging Output: Add debugging output to your script to print out the variables and the commands you're executing. This can help you identify any unexpected values or issues.

    Try Different Shell: In GitHub Actions, you might specify the shell to use in your workflow. If you're currently using bash, consider trying a different shell, such as sh, to see if that affects the behavior.

    Alternative Approach: Instead of using a pipe to pass the data to the security command, you could try saving the decoded certificate data to a temporary file and then passing the file to the security command.

    Update macOS Version: If you're using a specific version of macOS on GitHub Actions, consider updating the version to match the one you tested locally. Sometimes, minor differences in system libraries or utilities can lead to behavior discrepancies.

    GitHub Actions Environment Changes: Check if there have been any updates or changes to the GitHub Actions macOS runners that might be impacting the behavior of the security command.

    Contact GitHub Support: If the issue persists and you've exhausted your troubleshooting options, consider reaching out to GitHub Support for assistance. They might have insights into recent changes or known issues related to GitHub Actions environments.

As of my last update in September 2021, I don't have access to the most recent GitHub Actions environment changes, so it's important to consult the latest documentation and support resources for accurate guidance.

**Comment by mywristbands** [comment, marked as answer]
I managed to figure it out!

For some reason Process Substitution (using `<(command-goes-here)` to direct the output of a command to a temporary file) is not working correctly for me in GitHub Actions anymore, or at least not working anymore in my particular case. Though note that this still works just fine for me on my local Mac.

I found that one way to get around this is to instead (1) save the command output to a regular file, then (2) pass this filename into the import command, and then (3) delete the file afterwards. So here's my new code:
```
echo $SIGNING_CERTIFICATE_P12_DATA | base64 --decode > signingCertificate.p12
security import signingCertificate.p12 \
                -f pkcs12 \
                -k build.keychain \
                -P $SIGNING_CERTIFICATE_PASSWORD \
                -T /usr/bin/codesign
rm signingCertificate.p12
```

**Reply by zhiqingchen** [reply]
it works

**Comment by thankiyash** [comment]
For anyone else going through this error, if you have environment variables, make sure you specify the env on the file.

```
jobs:
  build_with_signing:
    runs-on: macos-latest
    
    ##This fixed it for me###
    environment: stage

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Install the Apple certificate and provisioning profile
```

**Reply by mablds** [reply]
it worked for me! I didnt know how to point to the correct environment secrets/variables

thanks!!!
